### PR TITLE
Improve reading facets from ESGF search results

### LINKED ISSUE #1873: Download from ESGF saves to wrong path if facet value contains dots
Example:
```python
from esmvalcore.esgf import find_files
find_files(project='obs4MIPs', short_name='pr', dataset='GPCP-V2.3')[0].local_file('/')
```
results in
```
PosixPath('/obs4MIPs/GPCP-V2/3/v20180519/pr_GPCP-SG_L3_v2.3_197901-201710.nc')
```
while it should be
```
PosixPath('/obs4MIPs/GPCP-V2.3/v20180519/pr_GPCP-SG_L3_v2.3_197901-201710.nc')
```

### PR DESCRIPTION
**Pull request by bouweandela**
## Description

Improve the way facets are read from the ESGF search results and use them to create the local directory structure. This results in more cases where the facets are available and fixes an issue with the obs4MIPs directory where a `.` in the dataset name was interpreted as a `/` when creating the local file path.

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

To try out the improved facets, run
```python
from esmvalcore.esgf import find_files

files = find_files(
    short_name='tas',
    mip='Amon',
    project='CMIP5',
    dataset='*',
    institute='*',
    ensemble='r1i1p1',
    exp='historical',
    timerange='2000/2000',
)

for file in files:
    print(file)
    print(file.facets)
```
and note that with this pull request all datasets have many facets.

Closes #1873

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Review by valeriupredoi — APPROVED** [review]
looking good, bud (although am no specialist on esgf searches but I do very much appreciate the amount of testing added here!) :beer:

**Comment by remi-kazeroni** [comment]
@ESMValGroup/tech-reviewers, would one of you have time to do a final check and merge this, please? It would be great to have this merged soon as it also helps resolving issues in #1609. Thanks 👍

File: esmvalcore/esgf/_download.py [hunk 1]
```diff
@@ -234,8 +235,41 @@ def same_file(result):
 
         return files
 
+    def _get_facets(self, results):
```
Comment (schlunma) [inline esmvalcore/esgf/_download.py:238]: Please add a docstring here

File: tests/integration/esgf/test_search_download.py [hunk 2]
```diff
@@ -115,14 +128,40 @@ def test_mock_search(variable, mocker):
...
 
-    with open(data_path / 'expected.yml') as file:
-        expected_files = yaml.safe_load(file)[json_file]
+    expected_results_file = data_path / 'expected.yml'
+    if expected_results_file.exists():
+        with expected_results_file.open(encoding='utf-8') as file:
+            expected_results = yaml.safe_load(file)
+    else:
+        expected_results = {}
+    print(expected_results)
```
Comment (schlunma) [inline tests/integration/esgf/test_search_download.py:137]:
```suggestion
```

**Review by schlunma — COMMENTED** [review]
Thanks for this @bouweandela, I have two additional minor comments on this.

Could you also please briefly summarize what the issue was and how this PR fixes that? To be honest, I don't really understand what's going on here, and you are adding a significant amount of code here. I think the linked issue (#1873) is only relevant for a small part of the changes.

**Comment by bouweandela** [comment]
@schlunma The main issue is that the facets on ESGF are often incorrect. I tried to work around this by reading them from the `dataset_id` (e.g. `CMIP6.CMIP.NCAR.CESM2.historical.r4i1p1f1.Amon.tas.gn.v20190308`) using the template `dataset_id_template_` that defines it (e.g. `%(mip_era)s.%(activity_drs)s.%(institution_id)s.%(source_id)s.%(experiment_id)s.%(member_id)s.%(table_id)s.%(variable_id)s.%(grid_label)s`). This is the code that is currently in the `main` branch.

Unfortunately, this does not work in many cases because either the `dataset_id_template_` is incorrect or there is a dot in the dataset name (often the case for obs4MIPs). This results in `ESGFFile`s with empty or very few facets. The new approach is to read the facets from the json response of the ESGF query and then try to correct them by reading them from the `dataset_id` (assuming at least that is correct). This is the problem that prevented you from using these facets to create datasets in the recipe `589.yml.txt` from this comment https://github.com/ESMValGroup/ESMValCore/pull/1609#issuecomment-1424395111.

If you run the code in the description of the pull request with the current `main` branch and then with this branch and look at the difference, you can see the improvement.

File: esmvalcore/esgf/_download.py (hunk 1 above)
Comment (bouweandela) [reply esmvalcore/esgf/_download.py:238]: Done in [9874227](https://github.com/ESMValGroup/ESMValCore/pull/1920/commits/9874227016311e992c5340b940547b0f1ab208b8)

File: tests/integration/esgf/test_search_download.py (hunk 2 above)
Comment (bouweandela) [reply tests/integration/esgf/test_search_download.py:137]: Why would you recommend removing the print statement? I found it quite helpful for debugging when I was writing the test.

**Comment by schlunma** [comment]
Nice, got it, thanks for the explanation @bouweandela! I tested the code from the PR description, and can confirm that the facets look much better with this PR!

Will merge after the tests have passed. Cheers!!

**Pull request merged**